### DISCUSSION TITLE
Show only N matches at end of file

### DISCUSSION THREAD

**Discussion by jungle-boogie** [BurntSushi/ripgrep#1888] [General] [answered]
Hello,

I would like to know if I can use `rg` to show the last N lines of a match.
Effectively something like

```
rg 'ripgrep is great' myfile.log | tail
```

I'm familiar with options `-A`, `-B`, and `-C` but I don't want `ripgrep is great` printed hundreds of times, only the last N times.

Is there an option with `rg` to achieve this?

Thanks!

**Comment by jungle-boogie** [comment]
Why do I want this?

```
2466 matches
2466 matched lines
1 files contained matches
1 files searched
277011 bytes printed
5170124 bytes searched
2.181747 seconds spent searching
2.186922 seconds
```

I don't care about 2455 of those matches; I only want to know the status of the last ten lines in my file.

**Comment by BurntSushi** [comment, marked as answer]
There is no way, no.

To answer the next natural question: no, I don't think I will add it. In order to know if a match is part of the last N matches, you have to know where the last match is. In order to know where the last match is, you need to search the entire input. Thus, by necessity, implementing this feature requires buffering the last N matches in memory and only printing them once the search has completed. ripgrep isn't really set up to do that.

Instead, I suggest using `tail`: `rg foo | tail -n 10` will print only the last ten matches.

**Reply by jungle-boogie** [reply]
Hi @BurntSushi,

Understood. I had a feeling your response would be to not add it into `rg`, because of the reasons you cited. Makes sense to me.

I asked the question because `rg` can already do so much, I didn't know if I was overlooking a one of the flags.

I'll use `tail` to accomplish reviewing last ten lines.

Thanks for rg and I await the next great release!

**Reply by SOF3** [reply]
for line-local patterns, "the entire input" is just one line, right? so the requirement would be just starting from the last line onward, or alternatively, batch the last B lines and check if there are N matches, continue with the next B lines if there are less than N matches

**Comment by krzentner** [comment]
For anyone who comes here still looking for a way to do this for several files, it can approximately be done on unix systems using `--pre`, `tac`, and `--max-count` like so (as long as you don't have a multi-line pattern and don't care about line numbers):

```rg --pre tac -m 10 <PATTERN>```

This will run `tac` on every input file, which maybe isn't ideal, but that can be worked around in exchange for running `rg` twice (usually faster):

```rg --pre tac -m 10 <PATTERN> $(rg -l <PATTERN>)```

or, if you have a very large number of matching files:

```rg -l <PATTERN> | xargs rg --pre tac -m 10 <PATTERN>```

If you do have a multi-line pattern or care about line numbers, you'll probably want to just use the shell:

```rg -l <PATTERN> | while read f; do rg <PATTERN> $f | tail -n 10; done```

Thanks for making such a great tool, @BurntSushi.
